### DISCUSSION TITLE
Q&A: EWC in practice — Fisher matrix estimation, lambda tuning, and forgetting detection

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#243] [Q&A]
## Questions on Elastic Weight Consolidation — Phase 6 Implementation

Phase 6 EWC work starts with Issue #241 (FisherMatrix ABC + EWCRegulariser). This Q&A thread collects implementation questions that came up during the design phase.

---

### Q1: How do you estimate the diagonal Fisher in practice without a full training dataset?

**Short answer**: Use a Monte Carlo approximation over the weight updates from the most recent `SLEEP_PHASE` consolidation pass.

The exact Fisher requires computing gradients over the full data distribution:

```
F_i = E[ (∂ log p(y|x, θ) / ∂θ_i)² ]
```

In an online system like ASI:BUILD we never have the full distribution. The approximation used in `InMemoryFisherStore` is:

1. Collect `WeightUpdate` events from the Blackboard for the last N LEARNING ticks (stored under `learning.weight_deltas`)
2. Treat each `delta` as a gradient sample
3. Diagonal estimate: `F_i ≈ mean(delta_i²)` across the sample set

This is the "empirical Fisher" — biased but tractable. For Phase 6.1 it's sufficient. A proper Fisher using log-likelihood gradients would require the loss function, which STDP kernels don't expose directly.

---

### Q2: How do you choose the EWC lambda (regularisation strength)?

There's no universal answer — it depends on the ratio of new-task learning rate to consolidation importance. Practical starting points:

| System characteristic | Recommended λ range |
|----------------------|---------------------|
| High plasticity (STDP A+=0.05) | 200–400 |
| Low plasticity (STDP A+=0.005) | 1000–5000 |
| Frequent SLEEP_PHASE cycles (< 100 ticks) | Lower λ (Fisher updates often) |
| Rare SLEEP_PHASE cycles (> 500 ticks) | Higher λ (consolidations are rare, protect aggressively) |

The default `EWCConfig.lambda_ = 400.0` in Issue #241 is tuned for STDPOnlineLearner with `A_plus=0.01, tau_plus=20ms` — the kernel defined in Issue #181.

**Tuning signal**: If the agent shows high `phase6_fisher_diagonal_norm` (Prometheus gauge) but low `phase5_weight_delta_norm` on new inputs, lambda may be too high — the EWC penalty is dominating learning. If old memories degrade after new LEARNING phases, lambda is too low.

---

### Q3: How do you detect that catastrophic forgetting is happening?

There's no single "forgetting metric" but several proxy signals in the Blackboard + Prometheus:

**Proxy 1 — Fisher norm drift**
```promql
phase6_fisher_diagonal_norm > 2 * avg_over_time(phase6_fisher_diagonal_norm[30m])
```
A sudden jump suggests the parameter space is shifting rapidly — potential forgetting risk.

**Proxy 2 — EpisodicEvent retrieval quality**
After each SLEEP_PHASE, run a spot-check: query Neo4j for 5 random `EpisodicEvent` nodes from > 10 cycles ago and ask the consciousness planner to reconstruct them. Track `episode_reconstruction_accuracy` as a Gauge.

**Proxy 3 — Weight delta magnitude spike**
```promql
rate(phase5_weight_delta_sum[5m]) > 10 * rate(phase5_weight_delta_sum[1h])
```
A 10× spike in weight delta rate often precedes forgetting — catch it before the next SLEEP_PHASE.

---

### Q4: Does EWC work with STDP? STDP updates are local (Hebbian), not gradient-based.

This is the most important question for Phase 6.

Classic EWC was designed for gradient-descent-based learners where the Fisher has a clear definition. STDP is a local Hebbian rule — there's no explicit loss function.

**Adaptation for ASI:BUILD**:
- Treat each STDP `WeightUpdate.delta` as a pseudo-gradient
- The "Fisher diagonal" is the **variance** of delta_i across recent ticks, not the squared gradient of a log-likelihood
- This becomes a **weight-variance regulariser**: parameters that change a lot (high variance) are penalised more aggressively

This is sometimes called "synaptic intelligence" (SI) rather than true EWC, but for the purpose of Phase 6.1, the interface (`FisherMatrixBase`, `EWCRegulariser`) is identical — only the `compute()` implementation differs.

If/when ASI:BUILD adds gradient-based learners in Phase 7, the same `EWCRegulariser` wrapper works with zero changes.

---

### Q5: Online EWC vs standard EWC — when does `EWCConfig.online=True` matter?

Standard EWC keeps only the Fisher from the **most recent** consolidated task. After Task B, it forgets the Fisher from Task A entirely.

Online EWC accumulates Fisher across tasks:
```
F_online = γ * F_previous + (1 - γ) * F_current
```

Set `EWCConfig.online=True, gamma=0.9` when:
- The agent encounters many sequential tasks (> 5 distinct domains)
- SLEEP_PHASE cycles are frequent (the Fisher updates often)
- Memory is sufficient to store a large accumulated Fisher diagonal

For Phase 6.1 with a single STDP kernel and a relatively stable episodic stream, `online=False` is correct — the single Fisher snapshot from the most recent SLEEP_PHASE is sufficient.

---

Refs: Issue #241 (EWC foundation), Discussion #240 (Phase 6 planning), Issue #186 (MemoryConsolidator), Issue #181 (OnlineLearnerBase / STDP)
